modules/packages/Sort.chpl

```diff
@@ -968,123 +968,196 @@ module MergeSort {
+                 start:int = Dom.low, end:int = Dom.high) {
 
     chpl_check_comparator(comparator, eltType);
 
     if Dom.rank != 1 {
       compilerError("quickSort() requires 1-D array");
     }
 
+    if Dom.stridable {
+      ref reindexed = Data.reindex(Dom.alignedLow..#Dom.size);
+      quickSort(reindexed, minlen, comparator);
+      return;
+    }
+
     // grab obvious indices
-    const lo = Dom.low,
-          hi = Dom.high,
+    const lo = start,
+          hi = end,
           mid = lo + (hi-lo+1)/2;
 
-    // base case -- use insertion sort
-    if (hi - lo < minlen) {
-      InsertionSort.insertionSort(Data, comparator=comparator);
+    if hi - lo < minlen {
+      // base case -- use insertion sort
+      InsertionSort.insertionSort(Data, comparator=comparator, lo, hi);
+      return;
+    } else if hi <= lo {
+      // nothing to sort
       return;
     }
 
-    // find pivot using median-of-3 method
-    if (chpl_compare(Data(mid), Data(lo), comparator) < 0) then
-      Data(mid) <=> Data(lo);
-    if (chpl_compare(Data(hi), Data(lo), comparator) < 0) then
-      Data(hi) <=> Data(lo);
-    if (chpl_compare(Data(hi), Data(mid), comparator) < 0) then
-      Data(hi) <=> Data(mid);
-
-    const pivotVal = Data(mid);
-    Data(mid) = Data(hi-1);
-    Data(hi-1) = pivotVal;
-    // end median-of-3 partitioning
+    // find pivot using median-of-3 method for small arrays
+    // and a "ninther" for bigger arrays
+    if hi - lo < 100 {
+      order3(Data, lo, mid, hi, comparator);
+    } else {
+      // assumes array size > 9 at the very least
 
-    var loptr = lo,
-        hiptr = hi-1;
-    while (loptr < hiptr) {
-      do { loptr += 1; } while (chpl_compare(Data(loptr), pivotVal, comparator) < 0);
-      do { hiptr -= 1; } while (chpl_compare(pivotVal, Data(hiptr), comparator) < 0);
-      if (loptr < hiptr) {
-        Data(loptr) <=> Data(hiptr);
-      }
+      // median of each group of 3
+      order3(Data, lo,    lo+1, lo+2,  comparator);
+      order3(Data, mid-1, mid,  mid+1, comparator);
+      order3(Data, hi-2,  hi-1, hi,    comparator);
+      // median of the medians
+      order3(Data, lo+1,  mid,  hi-1,  comparator);
     }
 
-    Data(hi-1) = Data(loptr);
-    Data(loptr) = pivotVal;
+    var (eqStart, eqEnd) = partition(Data, lo, mid, hi, comparator);
 
-    // TODO -- Get this cobegin working and tested
-    //  cobegin {
-      quickSort(Data[..loptr-1], minlen, comparator);  // could use unbounded ranges here
-      quickSort(Data[loptr+1..], minlen, comparator);
-    //  }
+    if hi-lo < 300 {
+      // stay sequential
+      quickSort(Data, minlen, comparator, lo, eqStart-1);
+      quickSort(Data, minlen, comparator, eqEnd+1, hi);
+    } else {
+      // do the subproblems in parallel
+      forall i in 1..2 {
+        if i == 1 then
+          quickSort(Data, minlen, comparator, lo, eqStart-1);
+        else
+          quickSort(Data, minlen, comparator, eqEnd+1, hi);
+      }
+    }
```

Comment:
How about just
```chapel
serial (hi - lo < 300) do forall i in 1..2 { ...
```

assuming performance holds up?

Reply:
That approach is significantly slower.

Reply:
Huh.  OOC, does it make sense that it's slower that way?  I don't have any intuition that it should be.

Reply:
I have no idea what you mean by OOC...

The `serial` part makes use of task local storage and "runs" the parallel functions. The alternative presents more opportunity for compiler optimization and doesn't use task-local storage.

Reply:
"Out Of Curiosity".

Huh, ok, thanks.